Sorted meters by identifier issue[#577]

ecluster opened the pull request:
# Description
Meters are now sorted by identifier rather than id/name. Added an algorithm that maps ownProp id to sorted indices. Then the sorted meters are retrieved and passed in as a single meter. 

Fixes #[577]

## Type of change
- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist
- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations
The meter page loads in each meter individually rather than loading an array of meters. If this changes in the future and meters are loaded in differently, this code needs to be revisited and fixed appropriately. 

## Additional context
../src/client/app/containers/meters/MeterViewContainer.tsx (file that takes care of sorting)

Demo of sorted meters:
<img width="762" alt="Screen Shot 2022-03-07 at 9 18 33 PM" src="https://user-images.githubusercontent.com/53790807/157171367-b5cdfcd5-6b47-407e-84cf-f823be78b53e.png">

Joseph-M-Newman committed:
Included sorting method for meters.

The current state of meters for a non-logged in/admin user will be displayed in a sorted order by identifier. (Needs to take in account which meters are displayable)

ecluster committed:
Improved sorting algorithm by saving state of id after sorting.

1. Meters are now sorted by identifiers
2. Sorting takes account of displayable and non-displayable meters

Joseph-M-Newman committed:
Added comments to new code

Added comments to new code involving the algorithm to map ownProps.id to the sorted array index.

huss commented on src/client/app/components/meters/MetersDetailComponent.tsx:
```diff
@@ -34,7 +34,7 @@ export default class MetersDetailComponent extends React.Component<MetersDetailP
 	public render() {
 		const loggedInAsAdmin = this.props.loggedInAsAdmin;
-
+		
```
Please remove the extra tabs at the start of the line

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
 import { isRoleAdmin } from '../../utils/hasPermissions';
-
+import _ = require('lodash');
+var indexArray = 0;
```
We should try to avoid global variables. I appreicate you want to have a value that stays between function calls but maybe there is another way?

There should be a blank line before the function.

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
+var indexArray = 0;
 function mapStateToProps(state: State, ownProps: { id: number }) {
-	let meter = JSON.parse(JSON.stringify(state.meters.byMeterID[ownProps.id]));
+
```
No blank line at stat of a function.

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
-	let meter = JSON.parse(JSON.stringify(state.meters.byMeterID[ownProps.id]));
+
+	const meters = state.meters.byMeterID;
+	let sortedMeters = _.sortBy(_.values(meters).map(meter => 
```
meter should have a type

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
+		enabled: meter.enabled
+
+	})), 'identifier');
+	// Algorithm developed to keep sorted Array of meters's index in line with the current passed prop id.
```
It is good there is a comment here but it would be better if it told how it works.

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
 	if (currentUser !== null) {
 		loggedInAsAdmin = isRoleAdmin(currentUser.role);
 	}
+	// Keeps track of sortedMeter index. Since page is loaded multiple times resets when count hits the limit.
```
Can you please explain how this works? I think this is called for each meter so when it has been called that many times it resets back to zero to start the process again.

huss commented on src/client/app/types/redux/meters.ts:
```diff
@@ -55,10 +55,19 @@ export interface MeterMetadata {
 	gps?: GPSPoint;
 }
 
+export interface MeterSortedByIdentifier {
```
I'm unclear on where this is used.

huss commented on src/client/app/types/redux/meters.ts:
```diff
@@ -55,10 +55,19 @@ export interface MeterMetadata {
 	[meterID: number]: MeterMetadata;
 }
 
+export interface MeterMetadataByIdentifier {
```
Is this used outside this file? If not, does it need the export?

huss commented on src/client/app/containers/meters/MeterViewContainer.tsx:
```diff
@@ -10,23 +10,44 @@ import { editMeterDetails } from '../../actions/meters';
 	}
+	// Keeps track of sortedMeter index. Since page is loaded multiple times resets when count hits the limit.
+	// Needs to be visited when meter page is redesigned. 
+	if (indexArray != sortedMeters.length - 1){
```
formatting off, e.g., "){"

huss commented:
First, thanks to @ecluster and @Joseph-M-Newman for doing this work and submitting their first PR! I've left some comments on specific lines of code. Here are some overall comments:

I have run into several places where the code is not doing what I would expect:

- If you are an admin on the meters page and click the "Enable", "Disable", "Show" or "Hide" buttons then the value displayed does not change. If you select a time zone then it is not displayed. After you  click "save meter updates" they do change.
- If you edit the meter identifier and save, its location on the list does not change. If you refresh the page it does move.
- If you log out and go to the meter page then meters the user should not see are displayed. Refresh fixes it.
- I believe that I have seen edits to the meters, then logging out do not show on the non-admin version of the page. Refreshing seems to fix. Once the page went white (nothing displayed) and only came back on refresh.

I'm not sure what is up but suspect that the state is not properly updated and that the mapping you have to sorted meters has an issue. It may be possible to fix this but I am concerned there may be a deeper issue here. I'm happy to discuss more.

Other general comments:

- It would be nice to have more comments. I know the current code does not but we are trying to be better on new code.
- The CI failed. Please see the issues indicated.

huss commented:
Can you please merge in development so you are up to date. You want to be sure it works with new versions of the software and changes to package versions (esp. react-redux).

Joseph-M-Newman committed:
Merge branch 'OpenEnergyDashboard:development' into development

Joseph-M-Newman committed:
Modal boiler upload

Modal upload to fork. Included is basic example code for the modal animation + implementation on the meters page

Joseph-M-Newman committed:
Modal Update and revert sorting of meter page

Modal button only available  if logged in as admin

huss commented:
I finally had some time to look at this - sorry it took a little while (as we previously discussed). I believe I have a fix and it works on my branch. I changed src/client/app/containers/meters/MetersDetailContainer.tsx so the mapStateToProps modifies what it returns for meters. The complete return is:

```
	return {
		loggedInAsAdmin,
		// The state has all the meter info so map to get the id and identifier then sort that by
		// identifier and finally map so only have id as expected. This causes
		// MetersDetailComponent.tsx to process them in this sorted order
		// and pass them on to MeterViewContainer as desired.
		meters: _.sortBy(_.values(state.meters.byMeterID)
			.map(meter => ({ value: meter.id, label: meter.identifier.trim() })), 'label')
			.map(meter => meter.value),
		unsavedChanges: Object.keys(state.meters.editedMeters).length > 0
	};
```

where the comment explains what was done. Given you are actively working to modify the meter page with a complete revamp, I propose that you try to incorporate this into that work to see if it works. If so, we can close this PR and the fix will be in the PR from that effort. Please let me know if you have any questions or issues.

huss commented:
This was rolled into PR #773 so this PR is being closed.
